### PR TITLE
First cut at generating CLI docs for docs site

### PR DESCRIPTION
**Pull request by oscarbatori**
Our CLI has help text, we also use that text to generate docs for our documentation site at (DoltHub docs)[dolthub.com/docs].

Generating docs and command line output from the same content entails three separate concerns:
1. the raw content
2. modifying the content so the rendering is correct (in this case shell output and Gatsby build of `.mdx` files, JSX version of `.md`)
3. IO for writing to console (CLI help text) and files (`.md` files for building docs)

The goal of this PR is to move us towards a clean separation of these three concerns using Go templates. Specifically
- [x] implement helper types and methods for modifying content to render correctly in CLI output or `.mdx`
- [x] modify the raw content to the new data structures and text format that can be templated
- [x] update markdown generation code to simply request a template, rather than building document manually

Future work will be to use templates for the CLI output.

### CODE REVIEW

**Comment by oscarbatori** [comment]
Closes #317

**Comment by oscarbatori** [comment]
There is a failure in a Windows bats test run that I can neither reproduce locally, nor infer the issue from the logs. Trying to figure it out, but this is ready for review.

File: go/cmd/dolt/cli/documentation_helper.go [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+import (
+	"bytes"
+	"fmt"
+	"strings"
+	"text/template"
+
+	"github.com/liquidata-inc/dolt/go/libraries/utils/argparser"
+)
+
+type CommandDocumentForMarkdown struct {
```
Comment (bheni) [inline go/cmd/dolt/cli/documentation_helper.go]: Comment on exported function.

https://github.com/golang/go/wiki/CodeReviewComments#doc-comments

In the format:

https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences

File: go/cmd/dolt/cli/documentation_helper.go [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+	Command             string
+	CommandAndShortDesc string
+	Synopsis            string
+	Description         string
+	Options             string
```
Comment (bheni) [inline go/cmd/dolt/cli/documentation_helper.go]: Comments on all these

File: go/cmd/dolt/cli/documentation_helper.go [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+
+## Description
+{{.Description}}
+
+## Options
+{{.Options}}
+
+`
+
+func (cmdDoc CommandDocumentation) CmdDocToMd() (string, error) {
```
Comment (bheni) [inline go/cmd/dolt/cli/documentation_helper.go]: Comment on exported

File: go/cmd/dolt/cli/documentation_helper.go [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+	return templBuffer.String(), nil
+}
+
+type CommandDocumentation struct {
+	CommandStr string
+	ShortDesc  string
+	LongDesc   string
+	Synopsis   []string
+	ArgParser  *argparser.ArgParser
+}
```
Comment (bheni) [inline go/cmd/dolt/cli/documentation_helper.go]: Comments on all exported

File: go/cmd/dolt/cli/documentation_helper.go [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+
+func (cmdDoc CommandDocumentation) CmdDocToCmdDocMd(options string) (CommandDocumentForMarkdown, error) {
```
Comment (bheni) [inline go/cmd/dolt/cli/documentation_helper.go]: Comment on exported

File: go/cmd/dolt/cli/documentation_helper.go [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+	return CommandDocumentForMarkdown{
+		Command:             cmdDoc.CommandStr,
+		CommandAndShortDesc: fmt.Sprintf("`%s` - %s\n\n", cmdDoc.CommandStr, cmdDoc.GetShortDesc()),
+		Synopsis:            transformSynopsisToHtml(cmdDoc.CommandStr, synopsis),
+		Description:         longDesc,
+		Options:             options,
+	}, nil
+}
+
+func GetCommandDocumentation(commandStr string, cmdDoc CommandDocumentationContent, argParser *argparser.ArgParser) CommandDocumentation {
```
Comment (bheni) [inline go/cmd/dolt/cli/documentation_helper.go]: Comment on exported

File: go/cmd/dolt/cli/documentation_helper.go [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+	return CommandDocumentation{
+		CommandStr: commandStr,
+		ShortDesc:  cmdDoc.ShortDesc,
+		LongDesc:   cmdDoc.LongDesc,
+		Synopsis:   cmdDoc.Synopsis,
+		ArgParser:  argParser,
+	}
+}
+
+func (cmdDoc CommandDocumentation) GetShortDesc() string {
```
Comment (bheni) [inline go/cmd/dolt/cli/documentation_helper.go]: Comment on exported

File: go/cmd/dolt/cli/documentation_helper.go [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+	return cmdDoc.ShortDesc
+}
+
+func (cmdDoc CommandDocumentation) GetLongDesc(format DocFormat) (string, error) {
```
Comment (bheni) [inline go/cmd/dolt/cli/documentation_helper.go]: Comment on exported

File: go/cmd/dolt/cli/documentation_helper.go [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+		return "", err
+	}
+	var templBuffer bytes.Buffer
+	if err := templ.Execute(&templBuffer, docFormat); err != nil {
+		return "", err
+	}
+	return templBuffer.String(), nil
+}
+
+func (cmdDoc CommandDocumentation) GetSynopsis(format DocFormat) ([]string, error) {
```
Comment (bheni) [inline go/cmd/dolt/cli/documentation_helper.go]: Comment on exported

File: go/cmd/dolt/cli/documentation_helper.go [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+	}
+
+	return lines, nil
+}
+
+type DocFormat struct {
+	LessThan      string
+	GreaterThan   string
+	EmphasisLeft  string
+	EmphasisRight string
```
Comment (bheni) [inline go/cmd/dolt/cli/documentation_helper.go]: Comment on all exported

File: go/cmd/dolt/cli/documentation_helper.go [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+}
+
+var MarkdownFormat = DocFormat{"`<", ">`", "`", "`"}
+var CliFormat = DocFormat{"<", ">", "<b>", "</b>"}
+var SynopsisMarkdownFormat = DocFormat{"&lt;", "&gt;", "`", "`"}
```
Comment (bheni) [inline go/cmd/dolt/cli/documentation_helper.go]: Comment on all exported items.

File: go/cmd/dolt/cli/documentation_helper.go [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+	}
+	ret := templBuffer.String()
+	return ret, nil
+}
+
+type Supported struct {
+	Abbreviation string
+	Name         string
+	Description  string
+}
```
Comment (bheni) [inline go/cmd/dolt/cli/documentation_helper.go]: Comment on all exported

File: go/cmd/dolt/commands/dump_docs.go [hunk 13]
```diff
@@ -152,3 +126,15 @@ func (cmd *DumpDocsCmd) dumpDocs(idxWr io.Writer, dEnv *env.DoltEnv, dirStr, cmd
 
 	return nil
 }
+
+func CreateMarkdown(fs filesys.Filesys, path string, cmdDoc cli.CommandDocumentation) error {
```
Comment (bheni) [inline go/cmd/dolt/commands/dump_docs.go]: Comment on exported

**Comment by oscarbatori** [comment]
@bheni I had planned to just nix the GitHub Wiki and redirect to our docs site, are you okay with that?

File: go/cmd/dolt/commands/pull.go [hunk 14]
```diff
@@ -17,24 +17,24 @@ package commands
...
-	"\n" +
-	"\nMore precisely, dolt pull runs dolt fetch with the given parameters and calls dolt merge to merge the retrieved " +
-	"branch heads into the current branch."
-var pullSynopsis = []string{
-	"<remote>",
+var pullDocs = cli.CommandDocumentationContent{
+	ShortDesc: "Fetch from and integrate with another repository or a local branch",
+	LongDesc: `Incorporates changes from a remote repository into the current branch. In its default mode, {{.EmphasisLeft}}dolt pull{{.EmphasisRight}} is shorthand for {{.EmphasisLeft}}dolt fetch{{.EmphasisRight}} followed by {{.EmphasisLeft}}dolt merge <remote>/<branch>{{.EmphasisRight}}.
+
+More precisely, dolt pull runs dolt fetch with the given parameters and calls dolt merge to merge the retrieved branch heads into the current branch.
```
Comment (mjesuele) [inline go/cmd/dolt/commands/pull.go]: `head`?

File: go/cmd/dolt/cli/documentation_helper.go [hunk 15]
```diff
@@ -0,0 +1,265 @@
...
+
+## Options
+{{.Options}}
+
+`
+
+func (cmdDoc CommandDocumentation) CmdDocToMd() (string, error) {
+
+	// Accumulate the options and args in a string
+	options := ""
```
[the remark this answers is not in the document]
Comment (oscarbatori) [reply go/cmd/dolt/cli/documentation_helper.go]: We are building the string iterating, so we'd have to reset it to empty before starting to iterate, which is perhaps a bit less clear?

**Comment by oscarbatori** [comment]
Thanks @bheni and @mjesuele for your feedback. Pushed a new version.

**Pull request merged**